### ISSUE TITLE
Option to output `nan` when optimize.curve_fit falls instead of raise RuntimeError.

### ISSUE THREAD

**Issue by GalagaBits** [scipy/scipy#18925]

**Comment by mdhaber** [comment]
I agree that `optimize.curve_fit` should return an `OptimizeResult` that indicates whether it is successful (like most other functions in `optimize`) rather than raising an error, although my reasons would be different. In the meantime, please use `try`/`except` to prevent one failure from stopping the other code.

**Comment by mdhaber** [comment]
Also, can you tell us more about your application?

- When you loop, are all the data sets that you're fitting a curve to the same size? 
- Are all the curves of the same family?
- How many parameters specify the curve?

**Comment by tupui** [comment]
Just a fly by comment about speed of try/except. IIRC, with Python 3.11, it's a no-op so using try/except is free 😃

**Comment by GalagaBits** [comment]
> Also, can you tell us more about your application?
> 
> * When you loop, are all the data sets the same size?
> * Are all the curves of the same family?
> * How many parameters specify the curve?
> 
> What I'm interested in is whether this is an example in which a vectorized optimizer would be useful.

Yes I'll love to tell you more about the application. 

- All of the data are the same size. I am actually fitting 2D Gaussians to each frame in an array using `optimize.curve_fit`. If there is a better way to accomplish this without compromising much speed, please let me know!
- I'm not sure what is meant by the curves being of the same family. Could you elaborate on that?
- There are 7 parameters being fit from the Gaussian function I am using.

**Comment by mdhaber** [comment]
If I'm counting correctly, a 2D Gaussian would have five independent parameters. Can you describe the function or share the equation?

**Comment by GalagaBits** [comment]
Okay here is the function that I am using, which I got from stack overflow: 
```
        # creates a gaussian function, from stack overflow
        # https://stackoverflow.com/questions/21566379/fitting-a-2d-gaussian-function-using-scipy-optimize-curve-fit-valueerror-and-m

        def gaussian_2d(xy, amplitude, xo, yo, sigma_x, sigma_y, theta, offset):
            x, y = xy
            xo = xo.astype(float)
            yo = yo.astype(float)
            a = (np.cos(theta)**2)/(2*sigma_x**2) + (np.sin(theta)**2)/(2*sigma_y**2)
            b = -(np.sin(2*theta))/(4*sigma_x**2) + (np.sin(2*theta))/(4*sigma_y**2)
            c = (np.sin(theta)**2)/(2*sigma_x**2) + (np.cos(theta)**2)/(2*sigma_y**2)
            g = amplitude*np.exp( - (a*((x-xo)**2) + 2*b*(x-xo)*(y-yo) 
                                    + c*((y-yo)**2))) + offset
            return g.ravel()
```

**Comment by GalagaBits** [comment]
I have not looked at `gaussian_,kde` yet!

**Comment by mattgebert** [comment]
> I agree that `optimize.curve_fit` should return an `OptimizeResult` that indicates whether it is successful (like most other functions in `optimize`) rather than raising an error, although my reasons would be different. In the meantime, please use `try`/`except` to prevent one failure from stopping the loop. The difference in speed is probably quite small:

Just a comment on this, I also think it may also be worth returning an OptimizeResult - I can't find/see any current way of tracking what the optimised parameters are after hitting a `maxfev` limit (i.e. number of iterations), when a RuntimeError is raised. It would be useful to get partial / incomplete results, which may at least allow for initialising a more optimal solution where there is very slow long-iteration convergence, or even allowing the user to see if the regression is converging towards their desired complex data shape.
